### PR TITLE
GH-41836: [Java] Fix an undefined symbol error when ARROW_S3=OFF

### PR DESCRIPTION
**Pull request by philo-he**
### Rationale for this change

Fix undefined symbol error reported at runtime after ARROW_S3=OFF is used in compiling arrow Java.

### Are these changes tested?

Tested in my local. Not sure whether we need to add some test.

### Are there any user-facing changes?

No.

* GitHub Issue: #41836

### CODE REVIEW

**Comment by vibhatha** [comment]
@github-actions crossbow submit -g java

File: java/dataset/src/main/cpp/jni_wrapper.cc [hunk 1]
```diff
@@ -27,7 +27,9 @@
 #include "arrow/dataset/file_base.h"
 #include "arrow/filesystem/localfs.h"
 #include "arrow/filesystem/path_util.h"
+#ifdef ARROW_S3
```
Comment (kou) [inline java/dataset/src/main/cpp/jni_wrapper.cc:30]: Could you confirm whether `arrow/util/config.h` is included here implicitly?

Comment (PHILO-HE) [reply java/dataset/src/main/cpp/jni_wrapper.cc:30]: @kou, thanks for your review! I didn't find this header is included implicitly.

Comment (kou) [reply java/dataset/src/main/cpp/jni_wrapper.cc:30]: OK. How about the following?

```diff
diff --git a/java/dataset/src/main/cpp/jni_wrapper.cc b/java/dataset/src/main/cpp/jni_wrapper.cc
index 19a43c8d2f..f1819aab0b 100644
--- a/java/dataset/src/main/cpp/jni_wrapper.cc
+++ b/java/dataset/src/main/cpp/jni_wrapper.cc
@@ -25,9 +25,8 @@
 #include "arrow/c/helpers.h"
 #include "arrow/dataset/api.h"
 #include "arrow/dataset/file_base.h"
-#include "arrow/filesystem/localfs.h"
+#include "arrow/filesystem/api.h"
 #include "arrow/filesystem/path_util.h"
-#include "arrow/filesystem/s3fs.h"
 #include "arrow/engine/substrait/util.h"
 #include "arrow/engine/substrait/serde.h"
 #include "arrow/engine/substrait/relation.h"
```

`arrow/filesystem/api.h` includes `arrow/util/config.h` explicitly and `ARROW_S3` must be defined if S3 is enabled with `arrow/filesystem/api.h`.

Comment (PHILO-HE) [reply java/dataset/src/main/cpp/jni_wrapper.cc:30]: @kou, by including [arrow/filesystem/api.h](https://github.com/apache/arrow/blob/main/cpp/src/arrow/filesystem/api.h), a few extra headers are directly or indirectly included, but they are not actually used here. Can we just keep the current change?

Comment (kou) [reply java/dataset/src/main/cpp/jni_wrapper.cc:30]: If we want to keep the current style, we need to include `arrow/util/config.h` explicitly before we use `ARROW_S3`.

Comment (PHILO-HE) [reply java/dataset/src/main/cpp/jni_wrapper.cc]: @kou, I just realized what you meant. The code was updated. Thanks!

**Comment by kou** [comment]
Could you test that this change doesn't break #36934? A reproduced program exists in the description of #36934.
(You need to enable S3 for this.)

**Comment by PHILO-HE** [comment]
> Could you test that this change doesn't break #36934? A reproduced program exists in the description of #36934. (You need to enable S3 for this.)

@kou, I just did a test with S3 enabled. This change doesn't break it.

**Review by kou — APPROVED** [review]
+1

Thanks for checking it.

**Pull request merged**